### DISCUSSION TITLE
Local Message Storage

### DISCUSSION THREAD

**Discussion by misterupkeep** [bonfire-xmpp/bonfire#50] [Ideas]
Storing these conversations in a format like this might be better in the long-term -- especially when you forget the fine details you had discussed just months ago.

Storing some messages locally is obviously necessary, at least those that are on screen. 

However, we would also like to have commodities like search. And while there are server-side XEPs that support queries (namely Message Archive Management), their queries are limited in features by the server. 

To get around this, we would have to store those messages locally and perform whatever our custom features on it are.

Other than the brutalisticly obvious feature side of things, old messages are of value to people -- sentimental or otherwise, relying on the server staying up and storing our messages (which may not always be the case) is living on borrowed time.

In short: we _need_ to store messages locally.

The question then, is how?

**Comment by lukediamond** [comment]
IndexedDB is probably the most feature-rich option that exists on every platform. It would be nice to completely abstract message storage though, and have an in-memory cache of messages (swap from DB) to speed up access/scrollback. In memory would just be JS objects more than likely, which would still beat IDB (both in raw lookup speed and the lack of a promise resolution). Then we could swap out IDB for a custom solution for the Node/Electron and native mobile targets, and integrate any server-side stuff that's of interest.

[XEP-0049](https://xmpp.org/extensions/xep-0049.html) and [XEP-0098](https://xmpp.org/extensions/xep-0098.html) could perhaps be used to store things like search acceleration structures between clients. The most ambitious idea would be to use XEPs like these and potentially new ones to store data in a very loosely-defined way on the server, bypassing things like MAM and shifting more control to the client implementations. This would very possibly be opening Pandora's Box, but may allow the state/format to 'grow' more organically with less reliance on server specifications/interop. Getting anything done at the protocol level is painful so we obviously want to minimize how much that happens.

**Reply by misterupkeep** [reply]
In that case, a very plain and abstract interface a la:
```js
$messages.get('username@domain.jid');
$messages.append(jid, messages);
```
would be what we want.

And the internals would implement a last-used cache, while simultaneously solidifying into IndexedDB. In that case, we would download all messages regardless of server XEP suppport, only utilising the best method to download them.

---

There are a couple of edge cases to think about:
- What if your chat message history with someone spans, say 2-3 years? Do you want to download all of that data? I think we should prompt the user on how long do they want to sync -- if their history is longer than 1-2 months, or ~10k message, something like that.
- Secondly, we should look into all the different file sharing XEPs and figure out how they interplay with all this. Image embeds are a high-priority commodity, and it's only natural to expect they remain in chat when you scroll back up. How (and where) do we store those images?

---

IndexedDB storing format used to be (and maybe still is?) in compressed blocks. Cold storage needs to support quick fuzzy searching. Hitherto, this has worked by generating 4 character prefix maps -- mapping prefix to a list of blocks. This would provide with quick rough searches, and results are then refined by unpacking and searching through those blocks.

We should also provide support for random (rw) access, especially for things such as message deletion and editing.

A nice commodity is tagged searching, at least on predefined tags: `has:link`, `is:image`, etc.
